### PR TITLE
Issue 76 custom view for oauth login

### PR DESCRIPTION
**Pull request by finyak4**
## Description
Added a custom OAuth login view for Google/GitHub that:
- Creates a new user if none exists
- Logs in existing users
- Issues authentication tokens

## Tests
- Login for existing users
- New user creation

### CODE REVIEW

File: users/models.py [hunk 1]
```diff
@@ -367,6 +367,10 @@ def get_all(cls):
             QuerySet: QuerySet of all users.
         """
         return cls.active_users.all()
+    
+    @property
+    def id(self):
+        return self.user_id
```
Comment (mehalyna) [inline users/models.py:373]: Document in the model-level docstring or comments any reason for this aliasing, as it may be non-obvious for future maintainers.

File: users/tests/test_oauth_login.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
+from django.test import TestCase
+from django.contrib.auth import get_user_model
+from rest_framework.test import APIClient
+from rest_framework_simplejwt.tokens import RefreshToken
+import requests_mock
+from ..models import UserRole
```
Comment (mehalyna) [inline users/tests/test_oauth_login.py:6]: Use an absolute import (from users.models import UserRole) to avoid import issues if the test structure changes.

File: users/tests/test_oauth_login.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+from django.urls import reverse
+
+User = get_user_model()
+
+class OAuthTokenObtainPairViewTests(TestCase):
+    """
+    Test suite for the extended OAuthTokenObtainPairView that handles both
+    password and OAuth (Google/GitHub) authentication.
+    """
```
Comment (mehalyna) [inline users/tests/test_oauth_login.py:15]: Expand the docstring to include what endpoints are tested, and any notable edge cases covered.

File: users/tests/test_oauth_login.py
Comment (mehalyna) [inline users/tests/test_oauth_login.py:1]: Add negative tests for malformed payloads (missing required fields, invalid types).
Consider adding coverage for refresh tokens, if your endpoint provides them.
Use subTest contexts for repetitive checks with different inputs.
Ensure all external requests are properly mocked and do not leak real HTTP calls.
Add tests for cases where the provider responds with unexpected data shapes.
Add checks for token expiration and refresh behavior, if applicable.
Add comments for any magic values or non-obvious test logic.

File: users/views.py [hunk 4]
```diff
@@ -13,7 +13,14 @@
 from rest_framework.permissions import AllowAny
 from rest_framework.throttling import AnonRateThrottle
 from rest_framework_simplejwt.views import TokenObtainPairView
+from .serializers import CustomTokenObtainPairSerializer
+import requests
 
+from django.contrib.auth import get_user_model
```
Comment (mehalyna) [inline users/views.py:19]: Move this import closer to the model definition for clarity and to follow Django conventions.

File: users/views.py [hunk 5]
```diff
@@ -13,7 +13,14 @@
...
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from rest_framework import status
```
Comment (mehalyna) [inline users/views.py:22]: If these are used only for the new OAuth view, consider grouping related imports for readability.

File: users/views.py [hunk 6]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
                 {'status': 'error', 'message': 'An error occurred during verification.'},
                 status=status.HTTP_500_INTERNAL_SERVER_ERROR
-            )
\ No newline at end of file
+            )
+
+
+User = get_user_model()
+def get_default_role():
+    return UserRole.objects.get(role="user")
```
Comment (mehalyna) [inline users/views.py:179]: Add error handling for missing "user" roles (e.g., DoesNotExist exception).
Consider caching the result if this function is called frequently.

File: users/views.py [hunk 7]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
+
+class OAuthTokenObtainPairView(TokenObtainPairView):
```
Comment (mehalyna) [inline users/views.py:181]: Add explicit permission_classes (e.g., AllowAny) to clarify public accessibility for login endpoints.
 Add throttle_classes if brute-force login prevention is needed.

File: users/views.py [hunk 8]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
+        """
+        if provider == 'google':
+            return self.handle_google_oauth(access_token)
+        elif provider == 'github':
+            return self.handle_github_oauth(access_token)
+        else:
+            return Response(
+                {"error": "Unsupported OAuth provider"}, 
+                status=status.HTTP_400_BAD_REQUEST
+            )
```
Comment (mehalyna) [inline users/views.py:248]: Normalize provider string (e.g., .lower().strip()) for robustness against user input.

File: users/views.py [hunk 9]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
+                "first_name": first_name,
+                "last_name": last_name,
+                "password": "",
+                "user_phone": "",
+                "title": "",
+                "role": get_default_role()
+            }
+        )
+
+        return self.generate_jwt_response(user)
```
Comment (mehalyna) [inline users/views.py:304]: Add a timeout to the requests.get call to Google's API to avoid hanging on slow responses.
Log failed authentication attempts and returned data for debugging.
In error responses, clarify whether the error is due to an invalid token or a network issue.

File: users/views.py [hunk 10]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
+        except requests.RequestException:
+            return Response(
+                {"error": "Invalid Google token"}, 
+                status=status.HTTP_400_BAD_REQUEST
+            )
+        
+        email = data.get("email")
+        if not email:
+            return Response(
+                {"error": "Email not provided"},
```
Comment (mehalyna) [inline users/views.py:285]: The error message says "Email not provided" - consider clarifying if it’s a Google API issue or user profile privacy setting.

File: users/views.py [hunk 11]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
+                }
+            )
+
+            return self.generate_jwt_response(user)
+            
+        except requests.RequestException:
+            return Response(
+                {"error": "Invalid GitHub token"}, 
+                status=status.HTTP_400_BAD_REQUEST
+            )
```
Comment (mehalyna) [inline users/views.py:386]: Add a timeout to both requests.get calls.
Log GitHub API errors for easier debugging.
Check for verified email status in the email selection logic.
Consider handling the case where "name" is empty or not in "first last" format.

File: users/views.py [hunk 12]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
+            if not user.first_name and defaults.get('first_name'):
+                update_fields['first_name'] = defaults['first_name']
+            if not user.last_name and defaults.get('last_name'):
+                update_fields['last_name'] = defaults['last_name']
+            
+            if update_fields:
+                User.objects.filter(pk=user.pk).update(**update_fields)
+                user.refresh_from_db()
+        
+        return user, created
```
Comment (mehalyna) [inline users/views.py:420]: Document what happens if a user already exists with a different role.
Add logging for user creation and updates for traceability.

File: users/views.py [hunk 13]
```diff
@@ -164,4 +171,279 @@ def get(self, request, user_id, token):
...
+                "id": user.user_id,
+                # "username": username,
+                "email": user.email,
+                "first_name": user.first_name,
+                "last_name": user.last_name,
+                "user_phone": user.user_phone,
+                "title": user.title,
+                "role": str(get_default_role())
+            }
+        })
```
Comment (mehalyna) [inline users/views.py:449]: Include a serializer for user data rather than hardcoding fields, for easier future changes.
Return the actual role field, not str(get_default_role()), to avoid confusion between the user's role and the default role.
If username is supported, uncomment and set it accordingly.

**Review by mehalyna — CHANGES_REQUESTED** [review]

**Review by heeylena — APPROVED** [review]
I like your implementation of OAuth login view. Good job!
